### ISSUE TITLE
Multiple layers sharing same datasource

### ISSUE THREAD

**Issue by digitalsadhu** [mediasuitenz/mappy#66]
Currently, if you use the same datasource details to create 2 layers (for example) both layers will pull from the same datasource and therefore unnecessarily get the same data 2x.

*Proposed solutions:*

1. Behind the scenes, keep track of datasources and reuse the same data where possible

2. Break out datasources into a separate config and then allow layers to define which datasource they use 

```js
{
  dataSources: [
    {
      name: 'example-api'
      type: 'singlePoll'
      url: '...'
    }
  ],
  layers: [
    {
      name: 'layer-1'
      dataSource: 'example-api',
      ...
    }
  ]
}
```

@markstuart @chillu @matt-in-a-hat @Resonance1584 @flashbackzoo @benmanu thoughts?

**Comment by matt-in-a-hat** [comment]
Would option 2 allow us to have 1 source that is filtered in 2 different ways to produce data for 2 layers?

**Comment by digitalsadhu** [comment]
yes, you could present the same set of data filtered differently to create a bunch of layers but the data would only be pulled from the server 1x

**Comment by markstuart** [comment]
My preference would be option 2. This give a bit more separation between the objects which is always desirable in my opinion.

Would the layers then subscribe to the data load event on the appropriate dataservice to refresh the layer?

**Comment by digitalsadhu** [comment]
yea, I think the layer would inherit whatever settings the dataSource describes but perhaps it would be desirable to provide an overide?

example of where layer-1 uses the default but layer-2 only wants to get the data 1x
```js
{
  dataSources: [
    {
      name: 'example-api'
      type: 'longPoll'
      url: '...'
    }
  ],
  layers: [
    {
      name: 'layer-1'
      dataSource: 'example-api',
      ...
    },
    {
      name: 'layer-2'
      dataSource: {
        name: 'example-api',
        type: 'SinglePoll'
      },
      ...
    }
  ]
}
```
Something like that? Perhaps the syntax could more clearly explain that the datasource defaults are being overwritten?

Maybe something like...
```js
{
  dataSource: {
    name: 'example-api',
    options: {
      type: 'singlePoll'
    }
  }
}
```

**Comment by markstuart** [comment]
Yep, I like the options object, makes it a little more obvious.

**Comment by chillu** [comment]
+1 on the last notation suggestion (named `dataSource` object with `options` object). Just out of curiosity on the use case: Are you displaying the samedata in different ways, e.g. markers vs. heatmap?

**Comment by digitalsadhu** [comment]
Yea we are using the layer relationships feature we added to display the same set of data as 2 layers. In layer one we filter the data down to just markers and layer 2 just polygons. Then we define a relationship between layer 1 and 2 and add a handler function that allows a user to click on a marker and have the appropriate polygon from layer 2 show up.

**Comment by markstuart** [comment]
Fixed in #78
